### PR TITLE
Improve video reader speed

### LINKED ISSUE #107: napari-dlc can not handle 200k+ frames videos during manual frame extraction
Hi DLC developer! I am recording video at 60 Hz for 1 h. I was not able to use napari to manually extract frames (freezes for minute every move of slide-bar). DLC=2.3.7, Ubuntu 22.04. The manual extraction GUI is working fine in DLC 2.1.8.2. Can you briefly confirm on your side with a large frame number video if it's OS related issue or napari itself. Thank you!
Ziyan

### LINKED ISSUE #209: Video reader remains slow for large videos
### Context

Our current VideoReader implementation is not optimal, and improvements to either the current code or using potentially faster readers such as decord2 may be helpful.

See:
- #89 
- #107 
- https://github.com/DeepLabCut/DeepLabCut/issues/3264:
> I just tested. napari-deeplabcut 0.3.0.0 is still slow, though better than the old version. There is still a noticeable delay (~10 seconds) between dragging the video progress bar and displaying the new frame.
> 
> My RAM, CPU, and GPU should be very powerful (64 GB DDR5 + Intel Core i7-14700 KF + RTX 4080 16GB). 

 _Originally posted by @ChaoqunYinNeuroscience in [#3264](https://github.com/DeepLabCut/DeepLabCut/issues/3264#issuecomment-4445820407)_

### PR DESCRIPTION
**Pull request by C-Achard**
### Scope

This pull request introduces significant improvements to the video reading logic

- Major refactor of the `Video` class and `read_video` function for faster video frame reads
  - Greatly improves both local frame seeking and fast slider updates
  - Dask chunk size is now memory-aware, and better suited for fast slider drags by default
- Fixes in visibility array shapes in tracking for query shape 1
- New regression tests to ensure correct behavior with transposed and singleton visibility shapes.

### Benchmarking

The changes bring improvements in two separate regards:

- Local seeking is now much faster as we avoid paying the decode cost on every frame read (even if they are within the same GOP)
- The Dask implementation used to create one delayed task per frame, which is suboptimal compared to proper chunking

The benchmark was done with an expensive-to-decode, 1000 frames GOP video `(keyint=1000:min-keyint=1000:scenecut=0:bframes=8:ref=4`, using `testrc2`), for 3 repeats (random: 50 frames, local: 5 windows of 10 frames).

95% CIs are shown (note that trials may not be fully independent due to caching)
image

### Automated summary

**Video reading improvements:**
- Refactored the `Video` class to better preserve decoder state, support efficient forward jumps, and provide new methods (`seek`, `read_next`, `grab_next`, `read_frame`, `read_frames`) for robust and performant frame access. This includes logic to avoid unnecessary seeks and optimize small forward jumps. (`src/napari_deeplabcut/core/io.py` [[1]](diffhunk://#diff-be6c6f73ee5c48122719f58a13394e1529cc5c4c7459e8bb7ff3977bb75c616bR864-R882) [[2]](diffhunk://#diff-be6c6f73ee5c48122719f58a13394e1529cc5c4c7459e8bb7ff3977bb75c616bL883-R996)
- Rewrote the `read_video` function to read video data in blocks (chunks) for improved efficiency, with automatic chunk size selection based on frame size. The new implementation now correctly uses Dask for block-wise reading and includes timing instrumentation for debugging. (`src/napari_deeplabcut/core/io.py` [src/napari_deeplabcut/core/io.pyL883-R996](diffhunk://#diff-be6c6f73ee5c48122719f58a13394e1529cc5c4c7459e8bb7ff3977bb75c616bL883-R996))

**Tracking visibility shape handling fixes:**
- Updated `expand_query_features_over_time` to robustly handle various visibility array shapes, including (K, T), (T,), (K,), and singleton batch/channel dimensions. This ensures compatibility with different model outputs and prevents shape mismatches. (`src/napari_deeplabcut/tracking/core/data.py` [src/napari_deeplabcut/tracking/core/data.pyL130-R153](diffhunk://#diff-026a366b7359b9b649a1e93ac27d32b9968225ec7e3e68766819f18d73f27580L130-R153))

**Testing and regression coverage:**
- Added new regression tests to verify that `expand_query_features_over_time` accepts both singleton and transposed visibility shapes, ensuring correct behavior for edge cases and preventing regressions. (`src/napari_deeplabcut/_tests/tracking/test_data.py` [src/napari_deeplabcut/_tests/tracking/test_data.pyR246-R314](diffhunk://#diff-5b94965e62e5cc5d6adf6a85ddbdd39f11e88ceeecafd52a71b4311c5f61b3f9R246-R314))

### CODE REVIEW

File: src/napari_deeplabcut/core/io.py [hunk 1]
```diff
@@ -74,6 +77,11 @@
 DLC_CANONICAL_H5_KEY = "df_with_missing"  # TODO use this key instead of str literal in all places
 FALLBACK_H5_KEYS = ["keypoints"]
 
+# -----------------------------------------------------------------------------
+#  Debug consts
+# -----------------------------------------------------------------------------
+LOG_VIDEO_READER_TIMING = False
+
```
[the remark this answers is not in the document]
Comment (C-Achard) [reply src/napari_deeplabcut/core/io.py:88]: This is on purpose to avoid flooding the debug log. This must be enabled specifically for video reading debug.

File: src/napari_deeplabcut/core/io.py [hunk 2]
```diff
@@ -880,43 +894,111 @@ def width(self):
...
+        self.seek(start)
+        out = np.empty((stop - start, self.height, self.width, 3), dtype=np.uint8)
+
+        for i in range(stop - start):
+            ok, frame = self.stream.read()
+            if not ok:
+                raise IndexError(f"Could not read frame {start + i}")
+            self._pos += 1
+            cv2.cvtColor(frame, cv2.COLOR_BGR2RGB, dst=out[i])
+
```
Comment (deruyter92) [inline src/napari_deeplabcut/core/io.py:952]: minor: can we call `read_next()` here? It seems to be slightly duplicated

File: src/napari_deeplabcut/core/io.py [hunk 3]
```diff
@@ -880,43 +894,111 @@ def width(self):
...
+    - Larger chunks: better for local frame access (going frame-by-frame),
+    but when changing chunks, feels slower. So dragging fast is more expensive.
+    - Smaller chunks: feels more responsive when dragging fast,
+    but chunks recompute more often, so more latency when dragging slowly or frame-by-frame.
+    """
+    bytes_per_frame = height * width * 3
+    target_bytes = target_mb * 1024 * 1024
+    chunk = target_bytes // bytes_per_frame
+    return int(min(max(chunk, 8), 32))
+
```
Comment (deruyter92) [inline src/napari_deeplabcut/core/io.py:985]: The number 8 seems a bit magic (sounds reasonable, but maybe should be configurable). Also on high-res, would this fixed floor risk over-allocation of memory?

File: src/napari_deeplabcut/core/io.py [hunk 4]
```diff
@@ -880,43 +894,111 @@ def width(self):
...
+def _choose_chunk_size(height: int, width: int, target_mb: int = 256) -> int:
+    """Here we have to trade off between:
 
-        def _read_frame(ind):
-            stream.set_to_frame(ind)
-            return stream.read_frame()
-
-        lazy_reader = delayed(_read_frame)
-    else:  # pragma: no cover
-        from pims import PyAVReaderIndexed
```
Comment (deruyter92) [inline src/napari_deeplabcut/core/io.py:909]: PyAV support is now removed, should be updated in the docstring I think

File: src/napari_deeplabcut/tracking/core/data.py [hunk 5]
```diff
@@ -127,12 +127,26 @@ def expand_query_features_over_time(
...
             vis = vis[..., 0]
-        elif vis.ndim == 3 and vis.shape[0] == 1:
-            vis = vis.squeeze(0)
+        elif vis.shape == (1, T, K):
+            vis = vis[0]
+        elif vis.shape == (1, T, K, 1):
+            vis = vis[0, ..., 0]
+        elif vis.shape == (K, T):
+            # Transposed time/query layout.
+            vis = vis.T
```
Comment (deruyter92) [inline src/napari_deeplabcut/tracking/core/data.py]: What if T == K, could this lead to unintended transpose?

File: src/napari_deeplabcut/_tests/test_reader.py [hunk 6]
```diff
@@ -319,8 +319,7 @@ def test_video_init_and_properties(video_path):
 def test_video_read_single_frame(video_path):
     """Check that we can read at least one frame correctly."""
     vid = Video(video_path)
-    vid.set_to_frame(0)
-    frame = vid.read_frame()
+    frame = vid.read_frame(0)
 
     assert isinstance(frame, np.ndarray)
     assert frame.shape == (50, 50, 3)
```
Comment (deruyter92) [inline src/napari_deeplabcut/_tests/test_reader.py:325]: Maybe some additional testing would be useful (suggested by cursor):

```suggestion
def test_video_read_frame_out_of_bounds(video_path):
    vid = Video(video_path)
    with pytest.raises(IndexError):
        vid.read_frame(5)
    with pytest.raises(IndexError):
        vid.read_frame(-1)
    vid.close()


def test_video_read_sequential_frames(video_path):
    vid = Video(video_path)
    f0 = vid.read_frame(0)
    f1 = vid.read_frame(1)
    assert f0.shape == f1.shape == (50, 50, 3)
    vid.close()


def test_video_read_frames_block(video_path):
    vid = Video(video_path)
    block = vid.read_frames(1, 4)  # frames 1, 2, 3
    assert block.shape == (3, 50, 50, 3)
    assert block.dtype == np.uint8
    assert block[0].shape == (50, 50, 3)
    vid.close()


def test_video_read_single_frame(video_path):
    """Check that we can read at least one frame correctly."""
    vid = Video(video_path)
    frame = vid.read_frame(0)

    assert isinstance(frame, np.ndarray)
    assert frame.shape == (50, 50, 3)
```

**Review by deruyter92 — APPROVED** [review]

**Comment by deruyter92** [comment]
@C-Achard great improvements, looks good! Also good that you made some figures. Mostly minor comments

File: src/napari_deeplabcut/tracking/core/data.py (hunk 5 above)
Comment (C-Achard) [reply src/napari_deeplabcut/tracking/core/data.py]: Tried to make it safer for ambiguous cases, let me know what you think!

File: src/napari_deeplabcut/core/io.py (hunk 4 above)
Comment (C-Achard) [reply src/napari_deeplabcut/core/io.py:909]: Thanks!

File: src/napari_deeplabcut/core/io.py (hunk 3 above)
Comment (C-Achard) [reply src/napari_deeplabcut/core/io.py:985]: I will make it configurable via an env var

Comment (C-Achard) [reply src/napari_deeplabcut/core/io.py:985]: > Would this fixed floor risk over-allocation of memory?

Yes, definitely. It can now be set using os

**Review by deruyter92 — APPROVED** [review]
Good solution! Very helpful PR, I think it's ready to merge!
